### PR TITLE
Automated test suite

### PR DESCRIPTION
**Pull request by spence-man**
Added configuration files and documentation to setup Automated Testing.

### CODE REVIEW

File: README.md
Comment (tatwell) [inline README.md]: Let's move this section to its own wiki page in repo and link to that from here.

File: tests/acceptance/WPFirstCest.php
Comment (tatwell) [inline tests/acceptance/WPFirstCest.php]: Two questions:

1. What does this name mean?
2. Why doesn't this test subclass some kind of test class? What is the point of all those `_support` classes if they wouldn't be used in a case like this?

File: tests/wpunit/UserTest.php
Comment (tatwell) [inline tests/wpunit/UserTest.php]: Please make this a real test. You could test creating a new user.

File: README.md
Comment (tatwell) [reply README.md]: What I would want to see here is how to run tests: all of them, a single test, or a group (if applicable). A simplified version of usage section.

**Review by tatwell — COMMENTED** [review]
I like what I see. I'm a little worried by all the boilerplate. If it can be ignored and exists to simplify actually writing tests, that's fine.

Also, can I see clear examples of unit, functional, and acceptance test?

I still need to pull and actually try to run tests. Will probably have additional feedback then.

File: README.md
Comment (tatwell) [inline README.md]: Tests are essential part of an application so move these install steps into main Installation section above.

File: README.md
Comment (tatwell) [inline README.md]: Can you add a composer config file to repository so user can simply run something like `composer install` and it will install required dependencies, as with npm or bundler?

File: README.md
Comment (tatwell) [inline README.md]: Can this be added to repository so test are ready to roll (or did you already)?

File: README.md
Comment (tatwell) [inline README.md]: Can these be set up with default values so tests are ready to run? All the special install steps are off-putting.

File: README.md
Comment (spence-man) [reply README.md]: Ok, I'll transfer the Troubleshooting section to the Wiki. I'll also group together a simplified version  of automated testing and send the rest to the wiki.

File: README.md
Comment (spence-man) [reply README.md]: There should be a `composer.json` file in the automated-test branch. It will install all the required dependencies with composer

File: README.md
Comment (spence-man) [reply README.md]: I wasn't sure if we wanted to hold the actual database on github. I'll add the file and remove that step.

File: README.md
Comment (spence-man) [reply README.md]: Ok, I'll take the special instructions. I just wanted  to make sure that your admin login  and password matches that. Since it isn't live yet we both probably have different passwords for each instance of the clone.  But I'll standarize the admin login so we use that same login info. (admin: chappress, password: password)

File: tests/acceptance/WPFirstCest.php
Comment (spence-man) [reply tests/acceptance/WPFirstCest.php]: That name is just a custom one when I did the installation suite. It's just a custom test to see if it works. The name is based off how I generated the file. You can generate tests on the command line and give them custom values.

File: README.md
Comment (spence-man) [reply README.md]: I'll change that syntax that you mentioned as well

File: tests/acceptance/WPFirstCest.php
Comment (tatwell) [reply tests/acceptance/WPFirstCest.php]: I see. I know it's hard because we don't have any real testable features yet. But I'd try to make it a real -- or realistic -- test. Something like:

```
# tests/acceptance/LoginTest.php
class UserLoginTest {
  // Tests
  public function expectsUserToLogin(AcceptanceTester $I) {
  }

  public function expectsAdminToLogin(AcceptanceTester $I) {
  }
}
```

Is the `WP` prefix necessary? I find it redundant. It looks like Codeception likes to suffix their text with `Cept`. I don't see the point with that either.

**Comment by tatwell** [comment]
@spence-man I glanced at changes. They look good. I'm still trying to run tests. Having some issues getting my system configured (unrelated to code here -- my problems). I'll work on that this morning.

**Comment by tatwell** [comment]
@spence-man I glanced at changes. They look good. I'm still trying to run tests. Having some issues getting my system configured (unrelated to code here -- my problems). I'll work on that this morning.

File: tests/acceptance/WPFirstCest.php
Comment (spence-man) [reply tests/acceptance/WPFirstCest.php]: Ok, I'll work on that. The  WP is not necessary. It was just a demo from the config. The name can be anything.  The user can generate the file with whatever name they want.

File: tests/wpunit/UserTest.php
Comment (spence-man) [reply tests/wpunit/UserTest.php]: I'll set that up as well. I didn't include at first because it needs it's own DB.  Apparently WPUnit loads, installs and configures fresh WordPress installation before each test method runs. I'll see what I can do.

File: tests/acceptance/WPFirstCest.php
Comment (spence-man) [reply tests/acceptance/WPFirstCest.php]: Regarding the Cept comment. Codeception supports a variety of ways to write a test. They use Cept(scenario-based format) and Cest (class based format). When the user generates a test, they specifiy (Cept,Cest,Unit) and that gets added on the end of the file name when they generate the test. I guess it's just a formatting they use. A user can generate a test on the command line with something like: `codecept g:cept Acceptance "TestFeature"`. It will generate the filename and the way the code is structured based on (cept,cest declaration).

 I think that's what the _support classes are for. They providing the scaffolding to create that file and to provide more customization if needed.

Either way, I'll try to narrow down all this into a user friendly format.

Comment (tatwell) [reply tests/acceptance/WPFirstCest.php]: Thanks for the explanation. If that's their convention, then let's follow it.

File: README.md
Comment (tatwell) [inline README.md]: Here can you just provide command to append the alias line to ~/.bash_profile?

    echo "alias codecept=./vendor/bin/codecept" >> ~/.bash_profile

Or however it's best done.

**Review by tatwell — COMMENTED** [review]
I finally got my system set up to run test and was able to run them. I got an error:

    [GuzzleHttp\Exception\ConnectException\ cURL error 7: Failed to connect to localhost port 8000...

This was in the `WPFirstCest` test.

Does it need to be running a local server? If so, this will need to be automated. I believe our Blogs app does this.

File: README.md
Comment (spence-man) [reply README.md]: Yes, needs to have local server running. 

Ok, I'll add that part as well.

Just a note:
I was going to add that instruction but I realized that each environment setup might be different. I use (oh-my-zsh) and my path & alias is in the .zshrc file. Other setups could use (.bashrc, .profile, .bash_login).  So hopefully the bash_profile will work.

File: README.md
Comment (tatwell) [inline README.md]: I'd like this changed, I don't like camelCase and in the previous example were using underscore_var_names. *Unless* you can show that camel-case is recommended PHP practice. (I'm not sure.)

In any case, let's the same user as for the dev database. Just add steps above to create a `chappress_test` database and give `chappress` user access to that.

Then update test config to use that database.

File: README.md
Comment (tatwell) [inline README.md]: This step needs to be automated. I don't want to have to start the server every time to run tests. It should also run on a different port than default dev server so there's no conflicts. That is, I can run tests while dev server is running.

We have this automated for tests in our Blogs app so feel free to reference that. I don't know that it's necessarily the best approach so feel free to research further.

**Review by tatwell — CHANGES_REQUESTED** [review]

File: README.md
Comment (spence-man) [reply README.md]: I created a bash script which is in the [wiki](https://github.com/chapmanu/chap-press/wiki#php-server-script). Or do you want something like 'rake serve' like it is in the blogs?

Comment (spence-man) [reply README.md]: I also created a bash script that enters all the mysql info. Maybe I can compile all this together for automating commands? Like 'deploy setup', 'deploy server' 

I'm also looking into another option. It's a composer package that might be helpful for this.

File: .gitignore
Comment (tatwell) [inline .gitignore]: debug log should not be in repo.

File: README.md
Comment (tatwell) [reply README.md]: Ok, I’ll check it out tomorrow.

My paradigm for running tests is Rails. Not the interface per se (it doesn't have to use rake). But something as simple and automated.

File: .gitignore
Comment (spence-man) [reply .gitignore]: ok just removed it. not sure why it was there because I had `*.log` in the gitignore file

Comment (tatwell) [reply .gitignore]: Ok that makes sense. I couldn't see the `*.log` entry in the PR.

File: README.md
Comment (tatwell) [inline README.md]: Have we got a fix for this yet? I need this step automated before I can approve.

File: README.md
Comment (tatwell) [inline README.md]: The info is outdated, yes?

File: README.md
Comment (spence-man) [reply README.md]: I find an interesting script. [Php server manager](https://github.com/cubny/php-built-in-server-manager)

File: README.md
Comment (spence-man) [reply README.md]: Ok, edited the wpTest to chapress_test. But you think we should just take the whole note off?

File: README.md
Comment (tatwell) [reply README.md]: I don't like this:

- https://github.com/cubny/php-built-in-server-manager#how-to-use-it

Little steps like this drive me crazy. Even if they do get bundled up into some kind of setup script.

Does codeception really not facilitate this in some way?

File: README.md
Comment (tatwell) [reply README.md]: Yes, good idea. Let's get this document as focussed as possible.

File: README.md
Comment (spence-man) [reply README.md]: So the way my setup is now. I just run the one server. I can do the test through Codeception. I can still login through my original set up. Codeception resets the database after each test, so the original isnt messed with.

Comment (spence-man) [reply README.md]: You shouldn't have to start the server every time. It just keeps running whether testing or not.

Comment (tatwell) [reply README.md]: Yes, this isn't acceptable. Unless it's absolutely necessary, in which case I'll create an issue and we can come back to it.

One of our Rails apps requires running `foreman start` before running tests. It's a minor nuisance but if it can be avoided it should be avoided. This is one of those [mise en place](https://en.wikipedia.org/wiki/Mise_en_place) things I'm insistent about.

Timebox the remainder of today to address this. If you can't crack it, open an issue and we'll move forward.

Comment (spence-man) [reply README.md]: Maybe that line is confusing and redudant

`**Run Local Server** at `/public` directory: `php -S localhost:8000``

I think I'll just remove that.

Comment (spence-man) [reply README.md]: Sorry, I think I misunderstood the original comment.

File: config/php.ini
Comment (tatwell) [inline config/php.ini]: This is interesting. Was this added by the new server plugin?

File: tests/_output/phpbuiltinserver.access_log.txt
Comment (tatwell) [inline tests/_output/phpbuiltinserver.access_log.txt]: These log files should be removed from repo and ignored in `.gitignore`.

File: tests/_output/phpbuiltinserver.errors.txt
Comment (tatwell) [inline tests/_output/phpbuiltinserver.errors.txt]: .gitignore

File: config/php.ini
Comment (spence-man) [reply config/php.ini]: I added that on. It was something I was testing for the previous story. The plugin just needed a pathfile for the .ini. So i created that directory. It looks like the Blogs site does something similar.  I just added some content that maybe would be useful down the line.

File: tests/_output/phpbuiltinserver.access_log.txt
Comment (spence-man) [reply tests/_output/phpbuiltinserver.access_log.txt]: ok, ill update that. looks like the extension added those.

**Review by tatwell — COMMENTED** [review]
So tests are running great. Nice work. *However*, now when I run the local server and I go to localhost:8000 in my browser, I get redirected to localhost:8080 which results in a 404.

Any idea why that's happening?

File: README.md
Comment (tatwell) [inline README.md]: Also, as much as I like a cool screenshot gif, I think it's takes up too much real estate and isn't really needed in the end. So let's remove it.

Comment (spence-man) [reply README.md]: No problem, just put in there initially. It does indeed take up space.

File: wpunit.suite.yml
Comment (tatwell) [inline wpunit.suite.yml]: Does the fact that port here differs from port in `codeception.yml` and `tests/wpunit.suite.yml` above matter?

Is there a single parent config setting that can be set somewhere so the extra ones can be removed?

File: README.md
Comment (spence-man) [reply README.md]: For the localhost error. I encountered that problem too. I’ve been troubleshooting that  for a bit. The default port for the extension is 8000. I tried to change it to 8080. Seems like its not interchangeable (saw the same port in the package files of that extension)

I think the solution is:

- leaving the extension port to 8000 
- use another port for our general server 
-  revert the previous commit

Sorry, my last commit where I added the extension seemed to have messed things up.

Comment (tatwell) [reply README.md]: Using a different port for local server is fine. I really like the way tests run now. I'm ok changing local server port. In fact, I'd prefer not to use 8000 as that is a common default and more likely to lead to conflicts.

File: wpunit.suite.yml
Comment (spence-man) [reply wpunit.suite.yml]: Not sure about the single parent config. The ports on both those files should be the same. 

Are you on the dev or master branch? It shows older DB name  'wpTest'.

Comment (tatwell) [reply wpunit.suite.yml]: My comment is on this pull request. Which makes me wonder: is this config file even being used? Tests all passed.

File: README.md
Comment (spence-man) [reply README.md]: Ok sounds good. I'll change that port.  I think I might have to revert to previous commit. I'll work on that.

File: wpunit.suite.yml
Comment (spence-man) [inline wpunit.suite.yml]: Not sure. When I checked that file on github it shows the updated info.

Comment (spence-man) [reply wpunit.suite.yml]: Oops sorry, that file is not supposed to be there. Looks like I uploaded that wpunit.suite.yml file to the dir root folder. Shouldn't be there. I'll remove that. The test/wpunit.suite.yml should be working properly.

Comment (spence-man) [inline wpunit.suite.yml]: That file is not doing anything. I'll remove that.

File: README.md
Comment (tatwell) [inline README.md]: Just ran tests and discovered this when they failed.

This shouldn't be necessary. `chappress_test` should be sufficient. Something is not right here.

File: README.md
Comment (tatwell) [inline README.md]: Reiterating comment on earlier changeset: remove this script and just list the php command here.

**Review by tatwell — CHANGES_REQUESTED** [review]

File: README.md
Comment (spence-man) [reply README.md]: The wpunit suite uses the wploader module. It configures and loads a blank WordPress. It states that this module needs a dedicated database. And this module cannot be used together with WPDb or DB module.  So I just created a separated database based on my interpretation of their docs.

Comment (spence-man) [reply README.md]: Seems that the test config file wasn't added. I just added that in. 

`cd ./tests/_wp_test_site`
`cp -v wp-config.php{-test,}`

Comment (tatwell) [reply README.md]: I don't understand the function or necessity of wpload module. Tests seemed to be working perfectly before this last change. Only problem was server wasn't running. Why was this needed?

Even if it is helpful, it should not require a separate database. The basic operation of tests should be:

- Call tests for command line
- Test runner sets up suite
- For each test:
  - Setup test (i.e. create test database tables)
  - Run test
  - Tear down test (i.e. truncate test database for next test)
- Test runner tears down test suite.

Comment (tatwell) [reply README.md]: Instead of a separate `-test` config file, can we modify the `-dist` version to switch settings based on environment. You can see an example here:

- https://github.com/chapmanu/blogs/blob/development/config/templates/wp-config-development.php.erb#L12

Trying to keep this as simple as possible. I feel like things are regressing a bit from where they were when I tested on Friday.

Comment (spence-man) [reply README.md]: Ok, I'll work on the config file.

Comment (spence-man) [reply README.md]: I encountered many problems after I added in the test port on fri. Tests were working but the site wouldn't load, or after editing, the site would load but the tests wouldn't. So l tried the latest setup to work with these issues. 

I'll look at the link and see I can make it work like that. 

So the setup I have, is to test on a demo site and the original public isn't compromised.  But from the link you sent, looks like it can be configured all from the same wordpress install.

 So we are testing the `/public` which will also be the live site?

Note for the wploader:
It's just an optional suite that was installed during the setup wizard. It can be removed if need be or added again later. It seems to combine  Codeception's unit testing with the Wordpress core testing suite.  

As noted in the docs: 

> The generated test case extends the WPTestCase class and it exposes all the methods defined by Codeception\Test\Unit test case and the Core suite \WP_UnitTestCase. 

> It has access WordPress defined functions and classes in integration tests; a wrapping of the WordPress PhpUnit based test suite provided in the WordPress repository

Comment (tatwell) [reply README.md]: That sounds like it's making helpers for testing WordPress core available. We shouldn't need that. Tests were working fine before it appeared.

I suspect the conflict between tests and dev server was caused by tests wiping out dev database because both were using same config file and referencing dev database.

Hopefully, a simple environment toggle in config file can solve this.

**Review by tatwell — COMMENTED** [review]
Pulled latest changes. Getting this error on acceptance tests:

> [GuzzleHttp\Exception\ConnectException] cURL error 7: Failed to connect to localhost port 8080: Connection refused (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)

**Comment by spence-man** [comment]
Not sure why the port is 8080. I have them set to 8222 for general and 8777 for the tests

**Comment by tatwell** [comment]
Server still redirects to port 8080 no matter what port I specify. 8080 refuses connection.

I try wiping out repo and starting from scratch.

**Comment by spence-man** [comment]
sounds like the database might be set to localhost:8080

**Comment by spence-man** [comment]
I think I got almost everything configure. All only thing is I'm trying to toggle the db in the config file.

I have this so far, but I'm not sure how to run this in codeception.

```
if (WP_ENV == 'test') {
  define('DB_NAME', 'chappress_test');
} else {
  define('DB_NAME', 'chappress_dev');
}
```

The key is making sure the WP_ENV var is set in the test environment. I'm not sure how Blogs manages it off the top of my head. I'd check the repo.

**Comment by tatwell** [comment]
FYI, I checked the out repo from scratch and followed README steps and still got same results:

- Acceptance tests fail.
- Dev server redirects to 8080 and refuses connection.

**Comment by spence-man** [comment]
I think it might be your database. You can try this.

```
cd tests/_data/
mysql -uroot chappress_dev < chappress_dev_dump.sql
```

**Comment by tatwell** [comment]
@spence-man 

That worked. This will be useful to keep in mind when we get to seed story:

    mysql -uroot chappress_dev < tests/_data/chappress_dev_dump.sql

It still redirects if I go to http://localhost:8222/. But if I add a path like, http://localhost:8222/foo, it loads. This could just be something screwy in the browser.

**Comment by spence-man** [comment]
Ya, happened to me as well. I think it's something with browse cache from previous tries. I just wiped my history or tried it in incognito mode and it was ok.

**Comment by tatwell** [comment]
If I revert to this changeset: https://github.com/chapmanu/chap-press/pull/2/commits/7b38e621a75eb3707d50c7b612ec953df8347158, tests all pass.

**Comment by spence-man** [comment]
Does that work even if the general port is not running?

I think my current config is good. I just have to tell codeception to use the config file with the db chappress_test somehow.

**Comment by tatwell** [comment]
Yes. No server running. I just run the tests. I don't know what's going on under the hood.

**Comment by spence-man** [comment]
I'm just afraid to go back because some configurations were messed up in previous commits. If I can solve this current problem, everything should be go to go.

**Review by tatwell — CHANGES_REQUESTED** [review]
I still can't approve this. I started with fresh clone and followed README steps.

- Dev server works. 👍 
- Tests fail.

If I simply run `codecept run`, the acceptance tests fail, though not like they were yesterday with cannot connect to server error. 

It I start dev server first, I get a different set of errors:

```
There were 3 failures:

---------
1) expectToActivatePluginCept: Log in as an admin and activate plugin
 Test  tests/acceptance/expectToActivatePluginCept.php
 Step  Click "Add New"
 Fail  Failed asserting that '/wp-login.php?redirect_to=http%3A%2F%2Flocalhost%3A8777%2Fwp-admin%2Fplugins.php&reauth=1' contains "/wp-admin/plugins.php".

Scenario Steps:

 3. $I->seeInCurrentUrl("/wp-admin/plugins.php") at tests/acceptance/expectToActivatePluginCept.php:16
 2. $I->amOnPluginsPage() at tests/acceptance/expectToActivatePluginCept.php:15
 1. $I->loginAsAdmin() at tests/acceptance/expectToActivatePluginCept.php:14


---------
2) expectUserAndAdminLoginCest: Login as an admin and add a new user
 Test  tests/acceptance/expectUserAndAdminLoginCest.php:expectsAddNewUser
 Step  Click "Add New"
 Fail  Form field by Label or CSS element with 'Username' was not found.

Scenario Steps:

 5. $I->fillField("Username","chappress-user") at tests/acceptance/expectUserAndAdminLoginCest.php:41
 4. $I->amOnPage("/wp-admin/user-new.php") at tests/acceptance/expectUserAndAdminLoginCest.php:40
 3. $I->loginAsAdmin() at tests/acceptance/expectUserAndAdminLoginCest.php:39
 2. // I am going to log in as Admin
 1. $I->amOnPage("/") at tests/acceptance/expectUserAndAdminLoginCest.php:13


---------
3) AddPostAsAdminCept: Add a post as an admin
 Test  tests/functional/AddPostAsAdminCept.php
 Step  Click "Add New"
 Fail  Link or Button by name or CSS or XPath element with 'Add New' was not found.

Scenario Steps:

 3. $I->click("Add New") at tests/functional/AddPostAsAdminCept.php:14
 2. $I->amOnPage("/wp-admin/edit.php") at tests/functional/AddPostAsAdminCept.php:13
 1. $I->loginAsAdmin() at tests/functional/AddPostAsAdminCept.php:12


FAILURES!
Tests: 9, Assertions: 6, Failures: 3.
```

Spencer, please clone the repo into a new directory, switch to this branch, and follow README steps. Then address the errors you see and/or update README as appropriate.

**Comment by spence-man** [comment]
I wiped my databases, cloned branch and ran through the install guide. Ran  localhost:8222 and the tests and no errors on my end. 

Can you run the tests as `codecept run --debug` and paste that over.

**Comment by spence-man** [comment]
You said you created empty database. Thats what I did.
But did you `Finish the final instructions through the WordPress admin installation.`?

Maybe your database was never generated during the install. I have encountered errors like that before and it was because there was no content in the db, (i.e. wordpress generated tables).

Seems like the tests can't find the content because it's not there.

**Comment by spence-man** [comment]
@tatwell Removed the relevant files. Tested unit and functional and it works on my end.

**Comment by tatwell** [comment]
Still one failure:

```
codecept run
Codeception PHP Testing Framework v2.3.4
Powered by PHPUnit 6.2.3 by Sebastian Bergmann and contributors.

Functional Tests (1) ----------------------------------------------------------------------------------------------------------------------------------------------------------------
✖ AddPostAsAdminCept: Add a post as an admin (0.52s)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Unit Tests (4) ----------------------------------------------------------------------------------------------------------------------------------------------------------------------
✔ DemoUnitTest: Reported interface (0.00s)
✔ VerifyAssertTest: Test_truth (0.00s)
✔ VerifyAssertTest: Test_string_contains_word (0.00s)
✔ VerifyAssertTest: Test_is_string (0.00s)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Time: 967 ms, Memory: 12.00MB

There was 1 failure:

---------
1) AddPostAsAdminCept: Add a post as an admin
 Test  tests/functional/AddPostAsAdminCept.php
 Step  Click "Add New"
 Fail  Link or Button by name or CSS or XPath element with 'Add New' was not found.

Scenario Steps:

 3. $I->click("Add New") at tests/functional/AddPostAsAdminCept.php:14
 2. $I->amOnPage("/wp-admin/edit.php") at tests/functional/AddPostAsAdminCept.php:13
 1. $I->loginAsAdmin() at tests/functional/AddPostAsAdminCept.php:12


FAILURES!
Tests: 5, Assertions: 5, Failures: 1.
```

**Comment by spence-man** [comment]
Not sure why that's happening.

**Comment by spence-man** [comment]
automated test suite configured with future adjustments

**Pull request merged**
